Generic dialogue rewrite

bubblobill opened the pull request:
### Identify the Bug or Feature request
closes #5606, #5505, #5147, .. and more I probably cannot find

### Description of the Change
Dialogues employ Jide Resizable.
Dialogue action buttons are now created as part of the dialog and not the content panel.
Buttons are added to the bottom of the pane in a scrollable button bar and should always be visible.
Buttons are laid out in order determined by OS (improved accessibility)
Button mnemonic keys are now separate i18n strings.
Content is placed within a scroll pane to ensure that oversize content can still be accessed.
Factory makes creating dialogues easier, and allows building in different steps, e.g.

```
private final GenericDialogFactory dialogFactory = GenericDialog.getFactory()
          .setDialogTitle(I18N.getText("action.addIconSelector"))
          .addButton(ButtonKind.CANCEL)
          .setCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
```
```
public AddResourceDialog() {
    super(new AddRessourcesDialogView().getRootComponent());
    setPreferredSize(new Dimension(550, 300));
    panelInit();
    dialogFactory
        .setContent(this)
        .addButton(
            ButtonKind.INSTALL,
            e -> {
              install = true;
              if (commit()) {
                dialogFactory.getDialog().closeDialog();
              }
            })
        .setDefaultButton(ButtonKind.INSTALL);
```

Still have to add missing I18n strings for buttons.

### Possible Drawbacks
Nightmare of a million dialogs might mean conversion isn't 100% yet.

### Documentation Notes
General dialog improvements

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/RPTools/maptool/5623)
<!-- Reviewable:end -->

bubblobill committed:
Changed to extending StandardDialog instead of JDialog.
Started factory methods.
Added a collection of standard buttons.

bubblobill committed:
Rewrite GenericDialog
Created GenericDialogFactory
Created ButtonKind, an enum of standard buttons for the button bar in
dialogs.

bubblobill committed:
Start of converting dialogs to new GenericDialog

bubblobill committed:
Start of converting dialogs to new GenericDialog

bubblobill committed:
Merge branch 'RPTools:develop' into GenericDialogue

bubblobill committed:
Tidying

bubblobill commented:
@kwvanderlinde pinging you as you tend to be my main dialog conflict source

bubblobill committed:
Try again. Conversion of existing dialogues.

bubblobill committed:
Try again. Conversion of existing dialogues. The ones I missed

bubblobill committed:
Reconnecting buttons with changed string keys.
Tweaks

kwvanderlinde commented on src/main/resources/net/rptools/maptool/language/i18n.properties:
```diff
@@ -163,7 +236,7 @@ Label.startup       = Startup
 Label.themes        = Themes
 Label.application   = Application
 Label.preferences   = Preferences
-Label.save          = Save
```
Need to keep this as it is still used in the **Preferences** dialog.

bubblobill replied:
It still exists. I just changed the key string to Button.save

kwvanderlinde replied:
Alright, you got me. Either way, **Preferences** won't open anymore.

bubblobill replied:
It looks like the refactor didn't take. This is why I normally avoid renaming i18n keys like the plague. I'll see if I can make it obey me.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/MapTool.java:
```diff
@@ -1335,6 +1335,7 @@ else if (AppPreferences.loadMruCampaignAtStart.get()) {
             AppActions.loadCampaign(campaignFile);
           }
         } catch (NoSuchElementException nse) {
+          AppPreferences.loadMruCampaignAtStart.set(false);
```
Just because there isn't a recent campaign this time, doesn't mean there won't be one next time. Let's not modify user preferences without their input.

bubblobill replied:
a) I invented "loadMruCampaignAtStart" so I should be able to do what I want with it.
b) This resulted from me trying to load a buggy campaign during development. What I actually want is to turn it off if the last time it tried it failed. I will make it a separate FR.

kwvanderlinde commented on src/main/resources/net/rptools/maptool/language/i18n.properties:
```diff
@@ -70,11 +70,116 @@ AutoSaveManager.status.autoSaveComplete = Autosave complete.  Elapsed time (ms):
+Button.next.mnemonic        = n
+Button.no                   = No
+Button.no.mnemonic          = n
+Button.ok                   = Okay
```
Keep this as "OK" as that is the norm for UIs.

bubblobill replied:
Whilst I disagree with you. You may find this amusing
![image](https://github.com/user-attachments/assets/365d836b-d405-4a2b-82fa-2a29f9e0757d)
Apparently the ButtonPanel class ignores me

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/startserverdialog/StartServerDialog.java:
```diff
@@ -60,16 +61,22 @@ public class StartServerDialog extends AbeillePanel<StartServerDialogPreferences
+            e -> {
+              // We don't have a good, server-side way of testing anymore.
+              boolean ok = MapTool.confirm("msg.info.server.networkingHelp");
+              if (ok) MapTool.showDocument(I18N.getString("msg.info.server.forumNFAQ_URL"));
```
Please add braces to this `if`.

bubblobill replied:
done

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/ui/startserverdialog/StartServerDialog.java:
```diff
@@ -226,17 +225,18 @@ protected void preModelBind() {
         .addActionListener(
             e -> {
-              if (getPortTextField().getText().length() == 0) {
+              if (getPortTextField().getText().isBlank()) {
```
Let's avoid the change in behaviour and use `.isEmpty()` instead. Same goes for the other replacements uses of `isBlank()` in this PR.

bubblobill replied:
isEmpty returns false for white-space.

A value of "  " is not useful. 

Unless I have blank and empty backwards.

kwvanderlinde replied:
You're right about their behaviours: `isEmpty()` is the same as `length() == 0` while `isBlank()` matches whitespace.

But that's my point. This change is tangential to the rest of the PR, and my gut says there's going to be some knock-on effects. I don't disagree that blanks for some of these fields are not particular useful, and aren't really handled well anyway.

bubblobill replied:
changed to isEmpty

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/swing/ButtonKind.java:
```diff
@@ -0,0 +1,112 @@
+    this.i18nMnemonicKey = i18nMnemonicKey;
+  }
+
+  public int getMnemonicKey() {
```
Might as well move this look up into the constructor so it's only calculated once and will be guaranteed to be consistent during the runtime.

bubblobill replied:
Good thinking.

kwvanderlinde commented on src/main/java/net/rptools/maptool/client/swing/ButtonKind.java:
```diff
@@ -0,0 +1,112 @@
+    if (keyString.length() == 1) {
+      return KeyEvent.getExtendedKeyCodeForChar(keyString.charAt(0));
+    } else {
+      // Literal strings, should not be translated.
```
Since these shouldn't be translated, let's hardcode them. Add an optional parameter to the constructor for the buttons that need it. If not provided, then try to look up a mnemonic from i18n.

bubblobill replied:
added a hard-coded map

bubblobill committed:
Merge branch 'RPTools:develop' into GenericDialogue

bubblobill committed:
Added i18n & hard-coded values to ButtonKind enum.
Increased preferredWidth to encourage disappearing of scrollbar.
Added context notes to i18n file
Fixed PreferencesDialogView.form refactor fail.
